File: src/api/schemas/fno.py
"""Pydantic response schemas for F&O API endpoints."""
from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict


class FNOCandidateResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    instrument_id: uuid.UUID
    symbol: str | None = None
    run_date: date
    phase: int
    passed_liquidity: bool | None = None
    atm_oi: int | None = None
    atm_spread_pct: Decimal | None = None
    avg_volume_5d: int | None = None
    news_score: Decimal | None = None
    sentiment_score: Decimal | None = None
    fii_dii_score: Decimal | None = None
    macro_align_score: Decimal | None = None
    convergence_score: Decimal | None = None
    composite_score: Decimal | None = None
    technical_pass: bool | None = None
    iv_regime: str | None = None
    oi_structure: str | None = None
    llm_thesis: str | None = None
    llm_decision: str | None = None
    config_version: str | None = None
    created_at: datetime


class IVHistoryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    instrument_id: uuid.UUID
    date: date
    atm_iv: Decimal
    iv_rank_52w: Decimal | None = None
    iv_percentile_52w: Decimal | None = None


class VIXTickResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    timestamp: datetime
    vix_value: Decimal
    regime: Literal["low", "neutral", "high"]


class FNOBanListResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    symbol: str
    ban_date: date
    is_active: bool


class PipelineTriggerResponse(BaseModel):
    status: str
    run_date: str
    phase1_passed: int = 0
    phase2_passed: int = 0
    phase3_proceed: int = 0


class ChainIssueResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    source: str
    instrument_id: uuid.UUID | None = None
    issue_type: str
    error_message: str
    raw_response: str | None = None
    detected_at: datetime | None = None
    github_issue_url: str | None = None
    resolved_at: datetime | None = None
    resolved_by: str | None = None


class ResolveIssueResponse(BaseModel):
    id: uuid.UUID
    resolved: bool
    source_health_status: str


class SourceHealthResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    source: str
    status: str
    consecutive_errors: int
    last_success_at: datetime | None = None
    last_error_at: datetime | None = None
    last_error: str | None = None
    updated_at: datetime | None = None

File: mobile/src/api/queries/fno.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import client from '../client';
import { STALE_TIMES } from '../../utils/constants';

export interface FNOCandidate {
  id: string;
  instrument_id: string;
  symbol: string | null;
  run_date: string;
  phase: number;
  passed_liquidity: boolean | null;
  atm_oi: number | null;
  atm_spread_pct: number | string | null;
  avg_volume_5d: number | null;
  news_score: number | string | null;
  sentiment_score: number | string | null;
  fii_dii_score: number | string | null;
  macro_align_score: number | string | null;
  convergence_score: number | string | null;
  composite_score: number | string | null;
  technical_pass: boolean | null;
  iv_regime: string | null;
  oi_structure: string | null;
  llm_thesis: string | null;
  llm_decision: string | null;
  config_version: string | null;
  created_at: string;
}

export interface IVHistory {
  instrument_id: string;
  date: string;
  atm_iv: number | string;
  iv_rank_52w: number | string | null;
  iv_percentile_52w: number | string | null;
}

export interface VIXTick {
  timestamp: string;
  vix_value: number | string;
  regime: 'low' | 'neutral' | 'high';
}

export interface FNOBan {
  symbol: string;
  ban_date: string;
  is_active: boolean;
}

export function useFNOCandidates(opts?: {
  run_date?: string;
  phase?: number;
  passed_only?: boolean;
  limit?: number;
}) {
  const params = new URLSearchParams();
  if (opts?.run_date) params.set('run_date', opts.run_date);
  if (opts?.phase) params.set('phase', String(opts.phase));
  if (opts?.passed_only) params.set('passed_only', 'true');
  params.set('limit', String(opts?.limit ?? 100));
  return useQuery<FNOCandidate[]>({
    queryKey: ['fno-candidates', opts ?? {}],
    queryFn: () => client.get(`/fno/candidates?${params.toString()}`).then((r) => r.data),
    staleTime: STALE_TIMES.reports,
  });
}

export function useVIXHistory(limit = 30) {
  return useQuery<VIXTick[]>({
    queryKey: ['vix', limit],
    queryFn: () => client.get(`/fno/vix?limit=${limit}`).then((r) => r.data),
    staleTime: STALE_TIMES.reports,
  });
}

export function useFNOBanList() {
  return useQuery<FNOBan[]>({
    queryKey: ['fno-ban-list'],
    queryFn: () => client.get('/fno/ban-list?active_only=true').then((r) => r.data),
    staleTime: STALE_TIMES.reports,
  });
}

export function useResolveChainIssue() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (issueId: string) =>
      client.post(`/fno/chain-issues/${issueId}/resolve?resolved_by=mobile`).then((r) => r.data),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['chain-issues'] });
      qc.invalidateQueries({ queryKey: ['source-health'] });
    },
  });
}
